apps/system/js/base_icon.js

```diff
@@ -0,0 +1,197 @@
+    if (ele) {
+      this.element = ele;
+      this.element.classList.add('active');
+      this.onrender && this.onrender();
+      if (this.UPDATE_ON_START) {
+        this.update();
+      }
+      this.debug('started early');
+    } else {
+      // XXX: To prevent icon ordering regression,
+      // we are waiting statusbar rendered and fetch the element here.
+      Service.request('Statusbar:getIcon', this.instanceID).then(
+        function(element) {
+          this.debug('get icon done.', element);
+          this.element = element;
+          this.onrender && this.onrender();
+          this.debug('started');
+        }.bind(this));
+    }
+  };
+  BaseIcon.prototype.start = function() {
+    if (this._started) {
+      return;
+    }
+    this.debug('starting..');
+    this._started = true;
+    this._settings = {};
+    if (this.constructor.REGISTERED_EVENTS) {
+      this.constructor.REGISTERED_EVENTS.forEach(function(evt) {
+        window.addEventListener(evt, this);
+      }, this);
+    }
+    if (this.constructor.OBSERVED_SETTINGS) {
+      this.constructor.OBSERVED_SETTINGS.forEach(function(key) {
+        Service.request('addObserver', key, this);
+      }, this);
+    }
+    this._start();
+    this.render();
+    // Integration test needs to toggle some icons actively.
+    Service.register('show', this);
+    Service.register('hide', this);
+    Service.registerState('isVisible', this);
+    Service.register('update', this);
+    // Integration test needs to stop some icons actively.
+    Service.register('stop', this);
+    if (this.UPDATE_ON_START) {
+      this.update();
+    }
+  };
+  BaseIcon.prototype.onrender = function() {};
+  BaseIcon.prototype._start = function() {};
+  BaseIcon.prototype.stop = function() {
+    if (!this._started) {
+      return;
+    }
+    this._started = false;
+    this._stop();
+    this.hide();
+    this.element && this.element.classList.remove('active');
+    Service.register('start', this);
+  };
+  BaseIcon.prototype._stop = function() {};
+  BaseIcon.prototype.isVisible = function() {
+    return !!(this.element && !this.element.hidden && this.enabled());
+  };
+  BaseIcon.prototype.handleEvent = function(evt) {
+    this.update();
+  };
+  /**
+   * this.update() will be controlled by the icon's manager actively.
+   * Override this function if your icon has more behavior than
+   * just showing/hiding.
+   */
+  BaseIcon.prototype.update = function() {
+    if (!this.enabled() || !this.element) {
+      this.debug('not started or no element');
+      this.debug(this._started);
+      this.debug(this.element);
+      return;
+    }
+    this.displayed() ? this.show() : this.hide();
+  };
+  BaseIcon.prototype.enabled = function() {
+    return this._started;
+  };
+  BaseIcon.prototype.updateLabel = function(type, active) {
+    if (!this.element && !this.isVisible()) {
+      return;
+    }
+    this.element.setAttribute('aria-label', navigator.mozL10n.get((active ?
```

Comment:
It would, if you change the string ID in the file from `id` to `id.ariaLabel` -- l10n.js will pick up l10nId change, read the `id` and update `aria-label` accordingly.